### DISCUSSION TITLE
PR or Fork of an o

### DISCUSSION THREAD

**Discussion by faanskit** [djansen1987/SAJeSolar#45] [Ideas]
Hi,

I've made an extensive overhaul of this integrations, basically starting from scratch. Only a few lines of code remains intact from original.

I'd like to know if you want this work to be pushed as a PR, or if I keep it a separate (and competing) integration.
It is **100% breaking the current system**, and no migration possible.

Today it only supports Inverters (since that is what I have), but it natively supports Multi-site, and it supports multiple inverters per site.

It is set-up via **Add integration**, so no more `configuration.yaml`. (Sorry for the Swedish in the screenshots, I also support translations as per standard HA framework)

![new_int](https://user-images.githubusercontent.com/10439868/201690221-765629b7-7eab-450b-b567-88b3d76bbd30.PNG)
![un_pass](https://user-images.githubusercontent.com/10439868/201690335-723a582e-0ef1-4344-93f1-c71ddc057a14.PNG)![sites](https://user-images.githubusercontent.com/10439868/201690252-3db3f7c4-ff15-45fb-bc30-e6d248171919.PNG)
![success](https://user-images.githubusercontent.com/10439868/201690367-2e250ba0-e689-43c9-b110-e39f7e3111a0.PNG)

I has **minimal amount of sensors**, but support more information as attributes. In this example, two sites:
![overview](https://user-images.githubusercontent.com/10439868/201690596-49330464-b727-410c-9639-6718c3780b4d.PNG)

Two sensors per site : **Status & Energy** 

![plant_status](https://user-images.githubusercontent.com/10439868/201690629-4a5153d1-4e52-4367-999c-2956e8c3bc03.PNG)![plant_el](https://user-images.githubusercontent.com/10439868/201690645-68560832-f5c4-482f-9d4e-16e12e0d3120.PNG)

Two sensors per inverter: **Energy and Power**

![inv_el](https://user-images.githubusercontent.com/10439868/201690695-a19d8bb4-6e20-43e9-a945-22798348fa38.PNG)![inv_pow](https://user-images.githubusercontent.com/10439868/201690707-df98a9da-3a38-4f4e-82c5-cf910344eb62.PNG)

As you can notice, under the **Power** sensor you find PV and Grid data.

I intend to add the following before I consider releasing it:
(1) Configuration of [multiple] sites (enable/disable)
(2) Configuration of inverter sensors (enable/disable); the inverter information is kind of redundant unless you need details
(3) Configuration of inverter PV/Grid attributes (enable/disable); the WEB query slows down the system quite a bit

Of course I'd like to support SEC and H1 - but then I'd need some more information about them and someone to test. In essence, I'd like to login to eSolar site with someone else credential while designing and building the system.

In a roadmap I consider to move the Esolar API code to become a PIP libary ay PyPi. Already now the integration code is broken out from the home assistant code, but for now it resides in a local file.

Should I also get intel about how to use the APP API's (encoding unknown) I will change from WEB API to APP API as they are more rich than the WEB API's.

What do you think. Would it be worth pushing this, or do I keep it separate?

/Marcus

**Comment by djansen1987** [comment]
HI @faanskit, cool that you have put in the work. First, I would like to warn all current users that if you switch, you will lose all history data in your energy dashboard when changing the consumption sensors!

Having said that, I should tell you that I do not have any solar panels myself and set up this integration helping out a family member who had got solar panels without HA support as the new dongles have no local API.

From there, I received some login details of other users who liked the integration, so I could reverse engineer the API to support their devices.
Having that said, I do not really care what is being added or improved to the integration as long as it stays working for my family member and other users. Adding UI support is something I wised to add for a long time as I also do not like the yaml setup, but I am not a python developer myself and have come this far by using my scripting knowledge in different programming languages and look at the integration of https://github.com/cyberjunky/home-assistant-toon_smartmeter to get me started.

So basically I am okay with you improving/maintaining this repository, but it has to be backward compatible.

**Reply by faanskit** [reply]
Thanks @djansen1987,

I 100% agree with your warning. Whatever I did will **break the current system** and **only supports inverters**, which is why I raised this thread. I did not want to push it as a PR before having a discussion as additional work would be required to support the current users (SEC, H1).

Neither am I a developer, merely a hobbyist coder that 1000-yrs ago (or so) used to be a code-:monkey: of dead languages. I just recently started to poke around in Python and having had a desire to understand the Home Assistant framework. The `ConfigFlow` in Home Assistant framework has been puzzling for me for quite some time, but I think I start to comprehend it now at least.

I originally wanted to switch to the APP API's as they appear better structured, but since i still miss some encoding parts I'm stuck with WEB API's reverse-engineered by you. I just moved them to use the `response` library.

For now, I think I'll keep it either a separate branch in my fork; or even set-up a completely parallel project. Given that I intend to push parts of it to PIP I think separate parallel project(s) is the way to go.

I'll give you a nudge when it's up so that you can steal back from me.

Thanks for the inspiration and work. For now I keep using your code for one of my systems to preserve history data, so I may have to figure out how to migrate data one day before I switch.

/Marcus

**Comment by RobbieDemaegdt** [comment]
I think it would be a good idea to just release this as a parallel integration because it's not backwards compatible.
I like the setup process, more robust than the yaml configuration.

So from my understanding there is no support yet for a battery?

**Reply by faanskit** [reply]
**Correct**, since I don't own a SEC (Load Monitor) or H1 (Hybrid Inverter) I simply cannot. Instead of a load monitor from SAJ, I make use of the HAN/P1 port of my energy meter to obtain load type of information.

I had to spend quite some time assessing the behavior of the Portal and the Web to define a Home Assistant model compatible with SAJ's information and data model. Ctrl-Shift-I is your friend in Chrome.

I decided to define a Home Assistant Device as a SAJ Plant/Site, and use Home Assistant Entities for the Inverters. This to get a bit more intuitive way to operate a multi-site installation in Home Assistant. I did consider to use the more "appropriate" way, to register each inverter as a Device. But, for a multi-inverter site - this became counter intuitive

The original integration (this integration) is not using the Device property at all, it's all a flat structure of entities.

Wrt. Hybrid Inverters, it would be very interesting to also understand if it can be controlled. If one can control via API, interesting capabilities can be done via Home Assistant, such as sell when spot-price is high and charge when spot-price is low. Nordpool data is readily available in Home Assistant already through an external integration.

But again, I don't have a Hybrid Inverter so I cannot even reverse engineer the portal. And the flat entity structure of current integration does not help understand.

**Reply by RobbieDemaegdt** [reply]
I might be able to help you with the Load Monitor as we have a battery installation at our home.
Would you want to implement it?

I also use the P1 port of my energy meter to obtain faster updates about the Import/Export of electricity.

**Reply by faanskit** [reply]
Without commitment, I can surely look into it. But I'm still confused about the Load Monitor (SEC) value add if you have a HAN/P1 meter.

From HAN/P1 - you obtain "sold" and "purchased" energy to/from the grid. 
From the SAJ inverter you obtain "produced".

AFAIKN, the load is the difference between them. purchased + produced - sold = load. This at least the set-up I have in Home Assistant. What additional information do you get from the SAJ SEC ?

Or, should I consider the SAJ SEC an alternative to the HAN/P1 data source?

Also, given the static nature of the electricity price in SAJ, cost and earnings is not even remotely correct if you have hourly rates.

These are the SAJ SEC entities. Can you pinpoint which you actually use that is not redundant from what you get from HAN/P1.

```
    sensors: saj_sec # Optional will only work with SAJ Sec Module
    resources:
      - nowPower
      - runningState
      - todayElectricity
      - monthElectricity
      - yearElectricity
      - totalElectricity # Energy -> Solar production
      - todayGridIncome
      - income
      - lastUploadTime
      - totalPlantTreeNum
      - totalReduceCo2
      - todayAlarmNum
      - status
      - plantuid
      - currency
      - address
      - isOnline
      - peakPower

      - pvElec
      - useElec
      - buyElec
      - sellElec
      - buyRate
      - sellRate
      - selfConsumedRate1
      - selfConsumedRate2
      - selfConsumedEnergy1
      - selfConsumedEnergy2
      - plantTreeNum
      - reduceCo2
      - totalGridPower
      - totalLoadPower
      - totalPvgenPower
      - totalPvEnergy
      - totalLoadEnergy # Energy -> Grid consumption
      - totalBuyEnergy
      - totalSellEnergy # Energy -> Return to grid
```

If you start with screenshots and an idéa how to model the Load Monitor wrt. what is a Device, what is an Entity and what should be an Attribute, then I can create something based on this integration.

But, to do a decent work - I would eventually need to log in to a system that has SEC and/or H1 and run my development integration against it.

**Reply by RobbieDemaegdt** [reply]
I'm using this in my energy dashboard.
      - chargeElec # Energy -> Home Battery Storage -> Energy going in to the battery (kWh)
      - dischargeElec # Energy -> Home Battery Storage -> Energy coming out of the battery (kWh)

and these are also important for me
      - batCapcity
      - batCurr
      - batEnergyPercent
      - batteryDirection
      - batteryPower
      
Apparently it's a H1 that I'm using, that is how my yaml looks like.

This is my yaml setup for the saj integration
- platform: saj_esolar
    username: *******@*******.***
    password: ******
    plant_id: 1
    sensors: h1
    resources:
      - nowPower
      - runningState
      - todayElectricity
      - monthElectricity
      - yearElectricity
      - totalElectricity # Energy -> Solar production
      - lastUploadTime
      - totalPlantTreeNum
      - totalReduceCo2
      - isAlarm
      - status
      - plantuid
      - currency
      - address
      - isOnline
      - devOnlineNum
      - selfUseRate
      - totalBuyElec # Energy -> Grid consumption
      - totalConsumpElec
      - totalSellElec # Energy -> Return to grid
      - chargeElec # Energy -> Home Battery Storage -> Energy going in to the battery (kWh)
      - dischargeElec # Energy -> Home Battery Storage -> Energy coming out of the battery (kWh)

      - pvElec
      - useElec
      - buyElec
      - sellElec
      - buyRate
      - sellRate
      - selfConsumedRate1
      - selfConsumedRate2
      - selfConsumedEnergy1
      - selfConsumedEnergy2
      - batCapcity
      - batCurr
      - batEnergyPercent
      - batteryDirection
      - batteryPower
      - gridDirection
      - gridPower
      - h1Online
      - outPutDirection
      - pvDirection
      - pvPower
      - solarPower
      - totalLoadPower

**Reply by faanskit** [reply]
> Apparently it's a H1 that I'm using, that is how my yaml looks like.

I almost guessed it, and it makes sense.

> I'm using this in my energy dashboard.
> - chargeElec # Energy -> Home Battery Storage -> Energy going in to the battery (kWh)
> - dischargeElec # Energy -> Home Battery Storage -> Energy coming out of the battery (kWh)
> 

These two should obviously be sensor entities. I'm thinking like this:

**Plant (2 x sensors per plant)**
|
+- Energy total (kWh - SensorStateClass.TOTAL_INCREASING)
|
+- Status (Online/Alarm/Offline)

**Inverter (2 x sensors per inverter)**
|
+- Energy total (kWh - SensorStateClass.TOTAL_INCREASING)
|
+- Power (W - SensorStateClass.MEASUREMENT)

**Battery (3 x sensors per Battry)**
|
+- Charge Energy (kWh - SensorStateClass.???)
|
+- Discharge Energy (kWh - SensorStateClass.???)
|
+- Power (W - SensorStateClass.MEASUREMENT)

Should a SensorStateClass be set? What type of value do they represent? Snapshot? Daily Total?
The following standard types are available, and the class can also be omitted.
```
class SensorStateClass(StrEnum):
    """State class for sensors."""

    MEASUREMENT = "measurement"
    """The state represents a measurement in present time."""

    TOTAL = "total"
    """The state represents a total amount.

    For example: net energy consumption"""

    TOTAL_INCREASING = "total_increasing"
    """The state represents a monotonically increasing total.

    For example: an amount of consumed gas"""

```

> and these are also important for me
> - batCapcity
> - batCurr
> - batEnergyPercent
> - batteryDirection

Can these be attributes? I want to eliminate as many sensors as possible. But since I don't own a battery system I don't know what is important.

If not, what SensorDeviceClass should they be? What units are they reported in the portal?

**Reply by faanskit** [reply]
One more thing: In the SAJ Portal, is it possible to control when to charge and discharge? A sniff from that API would be very interesting.

**Reply by RobbieDemaegdt** [reply]
The charge and discharge from the battery are a daily total.
![afbeelding](https://user-images.githubusercontent.com/5417599/201920730-0bf33efc-ce0d-4145-b32a-3920fc56130d.png)
![afbeelding](https://user-images.githubusercontent.com/5417599/201920786-6719d72d-12af-4760-815f-a2c7c999396a.png)

The battery percentage and so on can just be normal entities, I only use them on a dashboard.
I use this:
esolar batteryDirection, the value are numbers but should show Idle, Charging, Discharging
esolar batteryPower, is the power (in Watts) at the battery is charging or discharging
esolar batEnergyPercent, is the battery percentage (%)

Maybe a stupid question but what is the difference between a sensor and just an attribute?

There is no settings on my SAJ portal, I have an app on my phone that can switch the mode of the battery but the app is normally only to be used by installers but I found the credentials online.
But I heard something from my installer that they might be working on that feature.

**Reply by faanskit** [reply]
This picture illustrates what is a sensor, state and attributes:

![sensor](https://user-images.githubusercontent.com/10439868/201925496-26617141-9cde-42a6-9d97-e0ad8d470c8d.PNG)

Both sensor state and attributes can easily be used in jinja2 templates, which I assume you use for your dashboards.
e.g.:
```
{% set price = states('sensor.nordpool_kwh_se4_sek_3_095_0')%}
{% set average = state_attr('sensor.nordpool_kwh_se4_sek_3_095_0', 'average')%}
{{price}}
{{average}}
```
Based on your answer, both sensors should be `SensorStateClass.TOTAL_INCREASING`

Can you tell me the number<->text translation, e.g.:
```
0 = Idle
1 = Charging
2 = Discharging
? = Alarm?
```

**Comment by djansen1987** [comment]
If you like I can send you the json data from the end points I use for sec and H1. 

Ps @robshot what I offered to Marcus also goes for you, if you want to co-maintain this integration 
and take it to the next level be my guest. Only demand that I have is that is would stay backwards compatible.  This might me done through repairs (if this works for custom integrations). I also thought of merging it into the official saj integration of HA but Don't know which steps this would take. Don't feel this should be an HACS integration in the first time. But this might be out of my leak. 😀

**Reply by RobbieDemaegdt** [reply]
Yes I would like to help you guys out with maintaining the integration, only thing is my python is very rusty :p

I understand you want to keep it backwards compatible, no idea if we can fix it through repairs, not sure how that works.

It would make sense to merge this into HA but the only thing is that this API endpoint can break more easily than the endpoint on the local devices.
The strange part is that the saj integration in HA doesn't work for me, the local endpoints aren't working for some reason that is why I switch to the HACS version.
It would be nice if it would be merged and you got the option to choose between local api or remote api.

I'm not sure how private the json data is else you can attach in somewhere in the repo as development documentation.
if it's not possible you can send them to me so we have multiple people with that documentation if somebody of us needs it.

**Reply by faanskit** [reply]
Please do. I just set-up a temporary mail alias that you can send them to: apskaft-saj@usa.net

**Reply by faanskit** [reply]
As a side note, I have a LILYGO® TTGO T-CAN485 ESP32 CAN RS-485 on its way from China (Aliexpress). I intend to use it with ESPHome in Home Assistant and connect it to the inverters MODBUS port.

Hopefully I can completely ditch the API variant of I'm successful with this. I think Cloud Polling is wasteful.

**Reply by RobbieDemaegdt** [reply]
I also prefer local polling, first of all faster updates and second is not reliable on the remote API.
The AIO3 and wifi-d have terrible local API options, apparently the AIO3 should work but when I access the IP, I get nothing.
Also do they support MQTT to send data to the remote server but not when you change it to your local broker IP.

**Comment by RobbieDemaegdt** [comment]
Ok got it.
I think it would be nice to also have the graph for the battery percentage so I guess that means it needs to be a sensor?
![afbeelding](https://user-images.githubusercontent.com/5417599/201968605-9bc8e9a8-1335-4291-8d57-e993eafab9ec.png)

This is how it currently implemented.
![afbeelding](https://user-images.githubusercontent.com/5417599/201969254-7ae0fb17-c859-492e-afa3-71acb75c044e.png)

**Reply by faanskit** [reply]
Yes, it means a sensor. So, the minimal configuration would then be:
```
Battery (4 x Sensors per Battery)
|
+- Charge Energy (UnitOfEnergy.KILO_WATT_HOUR, SensorDeviceClass.ENERGY, SensorStateClass.TOTAL_INCREASING)
|     Attributes: TBD
|
+- Discharge Energy (UnitOfEnergy.KILO_WATT_HOUR, SensorDeviceClass.ENERGY, SensorStateClass.TOTAL_INCREASING)
|     Attributes: TBD
|
+- Power (UnitOfPower.WATT, SensorDeviceClass.POWER, SensorStateClass.MEASUREMENT)
|     Attributes: TBD
|
+- Battery Charge(PERCENTAGE, SensorDeviceClass.BATTERY, SensorStateClass.MEASUREMENT)
|     Attributes: TBD
```

Out of the remaining [current] sensors, would they be attributes and which sensors should the attributes be attached to?

A few screenshots of eSolar WEB would be nice to have, and JSON logs an absolute must.

/Marcus

**Reply by RobbieDemaegdt** [reply]
Ok I will provide the screenshots and json files you need.

From which page do you need the JSON data? getPlantDetailedInfo? Or is it from more realtimedata?

**Reply by faanskit** [reply]
Well, it's all about the relation to build an understanding.

As an example, 
- When I go the route Home->Plants->Inverters, the findDevicePageList returns a **single** list item.
- When I go the route Home->Inverter, the findDevicePageList returns **two** list items.

The difference is that in one request, the `plantuid` is present and in the other it is absent.

Since I don't have a battery system, I have no clue how this is visually represented in eSolar.  I am also curious if more than one battery can be installed, and how that is represented both in the Ui and especially in the data.

What I need is two things:
(1) The request (copy as fetch)
(2) The response (copy object)

E.g.
```
fetch("https://fop.saj-electric.com/saj/cloudMonitor/device/findDevicePageList", {
  "headers": {
    "accept": "application/json, text/javascript, */*; q=0.01",
    "accept-language": "sv-SE,sv;q=0.9,en-SE;q=0.8,en;q=0.7,en-US;q=0.6",
    "content-type": "application/x-www-form-urlencoded; charset=UTF-8",
    "sec-ch-ua": "\"Chromium\";v=\"106\", \"Google Chrome\";v=\"106\", \"Not;A=Brand\";v=\"99\"",
    "sec-ch-ua-mobile": "?0",
    "sec-ch-ua-platform": "\"Windows\"",
    "sec-fetch-dest": "empty",
    "sec-fetch-mode": "cors",
    "sec-fetch-site": "same-origin",
    "x-requested-with": "XMLHttpRequest"
  },
  "referrer": "https://fop.saj-electric.com/saj/cloudMonitor/device/deviceIframe?plantuid=XXXX&ifDialog=Y&ifIndex=N",
  "referrerPolicy": "strict-origin-when-cross-origin",
  "body": "officeId=XXXXX&pageNo=&pageSize=&orderName=1&orderType=2&plantuid=XXXX&deviceStatus=&localDate=2022-11-16&localMonth=2022-11",
  "method": "POST",
  "mode": "cors",
  "credentials": "include"
});
```
```
{
    "result": "OK",
    "page": "<ul>\n<li class=\"disabled\"><a href=\"javascript:;\">« Previous</a></li>\n<li class=\"active\"><a href=\"javascript:;\">1</a></li>\n<li class=\"disabled\"><a href=\"javascript:;\">Next »</a></li>\n<li class=\"disabled controls pageC\">Page <input type=\"text\" value=\"1\" onkeypress=\"var e=window.event||event;var c=e.keyCode||e.which;if(c==13)findDevicePageList(this.value,10,'');\" onclick=\"this.select();\"/> Total 1 Records</ul>\n<div style=\"clear:both;\"></div>",
    "list": [
        {
            "kitType": "N/A",
            "monthSellEnergyStr": "134.85",
            "todaySellEnergy": 0,
            "kitSn": "XXXX",
            "updateDateStr": "2022-11-15 23:20:04",
            "ifShowAFCI": 0,
            "powernower": "N/A",
            "mastermcufw": "V3.036",
            "type": 0,
            "devicetype": "R5-10K-T2",
            "onLineStr": "3",
            "displayfw": "V1.013",
            "devicepc": "XXXX",
            "powernow": 0,
            "isShowBattery": 1,
            "owner": "N/A",
            "todaySellEnergyStr": "N/A",
            "index": 0,
            "monthSellEnergy": 134.85,
            "devicesn": "XXXX",
            "userId": "XXXX",
            "onLine": 0,
            "slavemcufw": "V3.036",
            "isModuleExpire": 0,
            "totalSellEnergy": 7157.15,
            "totalSellEnergyStr": "7,157.15",
            "isShowHighVoltBat": 0,
            "isHistory": 0,
            "mark": 1,
            "plantName": "N/A",
            "dataTimeStr": "2022-11-15 16:20:00"
        }
    ]
}
```

I need the following endpoints:
- /monitor/site/getPlantDetailInfo
- /monitor/site/getPlantDetailChart2
- /cloudMonitor/device/findDevicePageList
- /cloudMonitor/deviceInfo/findRawdataPageList

And I need them from different paths, at minimum /findDevicePageList
E.g.
- Home->Plants->Inverters
- Home->Inverter 

Alternatively, provide me with credentials to this [mail](mailto:apskaft-saj@usa.net)

**Reply by RobbieDemaegdt** [reply]
I have send an email with some http request and responses.
I hope everything is there to continue the implementation.

I also included the battery realtime data and not only from the normal findrawdatapagelist

**Reply by faanskit** [reply]
Got it, could use a trace of getUserPlantList as well, just to ensure no major deviations.
Dumping some notes how about difference:
```
Solar Inverter R5     <->    Hybrid Inverter H1
===============================================
getPlantDetailInfo           getPlantDetailInfo
-----------------------------------------------
type: 0                      type: 3
income                       <missing>
todayGridIncome              <missing>
<missing>                    totalConsumpElec
<missing>                    totalBuyElec
<missing>                    totalSellElec
<missing>                    selfUseRate
todayAlarmNum                <missing>

getPlantDetailChart2
-----------------------------------------------
type: 0                      type: 1
pvType                       <missing>
userType                     <missing>
dayEnergy                    <missing>
<missing>                    viewBean{}
                               "pvElec": 0.21,
                               "useElec": 4.47,
                               "buyElec": 4.24,
                               "sellElec": 0.00,
                               "chargeElec": 0.00,
                               "dischargeElec": 0.02,
                               "buyRate": "94.85%",
                               "sellRate": "0%",
                               "selfConsumedRate1": "100%",
                               "selfConsumedRate2": "5.15%",
                               "selfConsumedEnergy1": 0.21,
                               "selfConsumedEnergy2": 0.23,
                               "plantTreeNum": 0.000,
                               "reduceCo2": 0.000,
                               "dataTime": 1668528000000

findDevicePageList
-----------------------------------------------
kitSn: "XXX"                kitSn: "N/A"
powernower: N/A             powernower: "218"
type: 0                     type: 2
devicetype: R5-15K-T2       devicetype:AS1-3KS-5.1
displayfw: N/A              displayfw:v3.100
todaySellEnergyStr: "N/A"   todaySellEnergyStr: 0.3



findBatteryRealTimeList
-----------------------------------------------
 N/A                   "list": [[{

                            "batVoltage": "51.1",
                            "batSOC": "15.0",
                            "dataTime": "2022-11-16 10:00:00",
                            "batCapcity": "100",
                            "batCurrent": "0.0",
                            "batPower": "0.0",
                            "batTemperature": "22.0",
                            "createDate": "2022-11-16 17:00:53"



findRawdataPageList
-----------------------------------------------
Same data structure, many differences.
PV Data = Same (Volt, Current)
Grid Data = Same (Volt, Current, Frequency)

deviceType: 0               deviceType: 2
pv3SeriesCurrent: N/A       pv3SeriesCurrent: 0.0-0.0-0.0-0.0
                            batEnergyPercent
                            backupTotalLoadPowerWatt
GRID Difference
                            r/s/t rOutPowerWatt
                            r/s/t rOutVolt
                            r/s/r rOutCurr
                            r/s/r rOutPowerVA
                            r/s/t rOutFreq
                            r/s/t rBackupPowerWatt
                            r/s/t rGridPowerWatt
                            r/s/t rOnGridOutFreq
                            r/s/t rGridPowerVA
                            r/s/t rOnGridOutVolt
                            r/s/t rOnGridOutPowerWatt                            
```


**Early observations**
1. The type parameter for **H1** is all over the place. `type=3`, `type=1`, `type=2` and `deviceType=2` depending on endpoint. For **R5** it is always `type=1`.
2. The request structure/order is the same for H1 and R5.
3. Unclear if content of `findBatteryRealTimeList` us useful, for now not in focus
4.  Some additional items for Site Sensor (totalConsumpElec, totalBuyElec, totalSellElec, selfUseRate )
5. Unclear if `getPlantDetailChart2` additional data should be used
6. Large amount of extra parameters for Inverter Power attributes (L1<->L3)

**Early question marks**
1. How would a multi-battery system represent itself? Does it even exist? 
2. How much of the additional data is actually useful?
3. getPlantDetailChart2 - additional things, can it be skipped?

**Early conclusions**
1. My (current) codebase will mostly work with H1 without major rewrites
2. Some complexity to be added to assess that darn type parameter
4. A bunch of additional detailed Grid parameters to be added as Attribues (for deviceType=2)
5. A few bat/backup parameters to be added (for deviceType=2)
6. findDevicePageList (type = 2), add displayfw

**Reply by RobbieDemaegdt** [reply]
I have send an email again with the request and response.

I have no idea how a multibattery setup would look like, no idea if it's even possible, I know you can extend the battery setup with another battery module but only till it has 3 extra modules attached.
To know this I will have to look at my inverter if there are multiple battery ports available and even then it's not 100% sure if there is no way to add a second battery.

From the getPlantDetailInfo is the totalbuyelec and the totalsellselec important as it needs to be used for the energy dashboard.
From the getPlantDetailChart2 is the chargeelec and dischargeelec important, again for the energy dashboard.
From the findBatteryRealTimeList is the battery soc important as this is the percentage of the battery. I see one thing that is missing and that is the battery direction which shows charging or discharging because the battery power is always positive.

**Comment by faanskit** [comment]
With some help from @robshot, progress have been made to also support H1.

![0_select](https://user-images.githubusercontent.com/10439868/202737218-aa9ef7b4-22b5-4814-bdbc-48be2008d753.PNG)
![1_success](https://user-images.githubusercontent.com/10439868/202737221-25463312-d7cf-47eb-81cd-04047337cc3d.PNG)
![2_configure](https://user-images.githubusercontent.com/10439868/202737225-619b8642-de94-416f-a468-f818fc814626.PNG)
![New_look](https://user-images.githubusercontent.com/10439868/202738031-b19a3dd8-d76e-49c9-a304-5bf094990c03.PNG)
![13_new_h1_attributes](https://user-images.githubusercontent.com/10439868/202738050-2fa870e3-c84b-4dfb-af62-6b8fb277970a.PNG)


A few more attributes to add, then it should be possible to test it on a live system. For now, simulated environment based on sniffed data for H1. R5 have been tested with my own plants.

**Comment by djansen1987** [comment]
@faanskit are you willing to share your code so i might be able to learn how you setup the UI integration ?

**Reply by faanskit** [reply]
Yes, I'll create a new repository on github soon.  Just need to alpha test H1. 
Right now on a business, so in a week or two...

**Reply by faanskit** [reply]
I've invited you to a first pre-alpha upload

**Comment by faanskit** [comment]
@robshot and @djansen1987, you are both added as collaborators to a new private repository.
Minimal instructions. Not yet tested as custom integration, only as "native" integration in HA development environment (Docker).
Supports R5 and H1, multi systems, natively.
Tested with multi R5 system live, and R5+H1 in basic test.

Add as custom_component. Settings->Integrations->Add integration-->eSolar

Insert your credentials and sensors shall be populated. To get more sensors and additional data as attributes_
Settings->Integrations->eSolar->Configure
Enable "Show inverter sensors" and "Show Photovoltaics and Grid attributes"

Have fun....

Limited support for a while; busy days ahead.

**Reply by RobbieDemaegdt** [reply]
Thanks, will try it out when I have some time.
Also I have a new setup, you know I had that battery issue so they replaced the old one with a new one so now I have 2 inverters on my H1 system for now untill they remove the first one (in the software) because it's not in use anymore.
Will see what it will do with the new integration.

**Comment by faanskit** [comment]
It's now in the open, tested with a Dual Plant R5 system and a Dual Plant R5+H1 system. No support for SEC.
https://github.com/faanskit/ha-esolar

**Reply by djansen1987** [reply]
👏

**Reply by faanskit** [reply]
I think there is a way to migrate the data from your integration into this integration. I have som SQL mongering to do, but it should be fairly straight forward.

I am considering to implement such a merge as a service, but I am hesitant to offer it programatically. Instead, I may consider an instruction.

### NOTE: THE BELOW HAVE NEVER BEEN TESTED - STEP 3 MAY RUIN YOUR SYSTEM

E.g.
**Step 1 - Find ID of current sensor**
```
SELECT id
FROM statistics_meta 
WHERE statistic_id= 'sensor.esolar_totalelectricity' 

Result
======
id
--
84
```

**Step 2 - Find ID of new sensor**
```
SELECT id
FROM statistics_meta 
WHERE statistic_id= 'sensor.Plant_XX_Total_Energy' 

Result
======
id
--
99
```

**Step 3A - Migrate Statistics**
```
UPDATE statistics
SET metadata_id = 99
WHERE metadata_id = 84
```

**Step 3B - Migrate Statistics Short Term**
```
UPDATE statistics_short_term
SET metadata_id = 99
WHERE metadata_id = 84
```

**Reply by djansen1987** [reply]
Your doing great work! And it it clear that the time you are putting in to it, i will never have the time and motivation for this as i do not have an SAJ module my self and the person i made it for is happy with the current setup. If you find a simple to use way to merge i am fine with you taking over the project

**Reply by RobbieDemaegdt** [reply]
Ok great to see this, I might try this on a test setup before doing this on my production HA.

**Reply by faanskit** [reply]
I've read somewhere that the core need to be stopped before this is performed. Have no clue how to stop the core.
Maybe best to take a snapshot of a virtual environment prior to running this...

**Reply by djansen1987** [reply]
HI Rob, maybe you can write a manual along the way for if other users want to switch to the intergration of faanskit?

**Reply by RobbieDemaegdt** [reply]
I still have to look into this, I'm very busy at the moment, also my GF don't want me to break stuff :p I might wanna testy this first on my test environment.
